### PR TITLE
Add support for conditional functions: IF, IFNULL, NULLIF and COALESCE.

### PR DESCRIPTION
**Pull request by petermattis**
Conditional functions differ from regular functions in that they do not
necessarily evaluate all of their arguments.

### CODE REVIEW

**Comment by jusongchen** [comment]
@pmattis, just a reminder that COALESCE and NULLIF are ANSI functions while
IF and IFNULL appears to be mySQL dialect.

On Wed, Sep 23, 2015 at 7:01 AM, Peter Mattis <notifications@github.com>
wrote:

> Conditional functions differ from regular functions in that they do not
> necessarily evaluate all of their arguments.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cockroachdb/cockroach/pull/2634
> Commit Summary
>
>    - Add support for conditional functions: IF, IFNULL, NULLIF and
>    COALESCE.
>
> File Changes
>
>    - *M* sql/parser/eval.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-0> (56)
>    - *M* sql/parser/eval_test.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-1> (13)
>    - *M* sql/parser/expr.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-2>
>    (102)
>    - *M* sql/parser/keywords.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-3> (1)
>    - *M* sql/parser/normalize.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-4> (11)
>    - *M* sql/parser/parse_test.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-5> (4)
>    - *M* sql/parser/sql.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-6>
>    (5678)
>    - *M* sql/parser/sql.y
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-7> (23)
>    - *M* sql/parser/type_check.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-8> (60)
>    - *M* sql/parser/type_check_test.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-9> (18)
>    - *M* sql/parser/walk.go
>    <https://github.com/cockroachdb/cockroach/pull/2634/files#diff-10> (5)
>
> Patch Links:
>
>    - https://github.com/cockroachdb/cockroach/pull/2634.patch
>    - https://github.com/cockroachdb/cockroach/pull/2634.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2634>.
>



-- 
Jusong Chen
408 315 2288

**Comment by petermattis** [comment]
@jusongchen Thanks for the reminder.

File: sql/parser/type_check.go [hunk 1]
```diff
@@ -228,6 +231,63 @@ func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum,
 		dummyLeft.Type(), op, dummyRight.Type())
 }
 
+func typeCheckConditionalExpr(expr *ConditionalExpr) (Datum, error) {
+	switch expr.Operator {
+	case If:
+		if len(expr.Exprs) != 3 {
```
Comment (tamird) [inline sql/parser/type_check.go]: we can do these checks statically. let's make ConditionalExpr an interface and give each of these a type

File: sql/parser/eval.go [hunk 2]
```diff
@@ -720,6 +723,59 @@ func (ctx EvalContext) evalIsOfTypeExpr(expr *IsOfTypeExpr) (Datum, error) {
...
+				return DNull, err
+			}
+			if d != DNull {
+				return d, nil
+			}
+		}
+		return DNull, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported conditional operator: %s", expr.Operator)
```
Comment (tamird) [inline sql/parser/eval.go]: this indicates a coding error, so should be util.Errorf.

File: sql/parser/type_check.go [hunk 3]
```diff
@@ -228,6 +231,63 @@ func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum,
...
+			if dummyArg == nil || dummyArg == DNull {
+				dummyArg = arg
+			} else if dummyArg != arg && arg != DNull {
+				return nil, fmt.Errorf("incompatible %s expressions %s, %s", expr.Operator, dummyArg.Type(), arg.Type())
+			}
+		}
+		return dummyArg, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported conditional operator: %s", expr.Operator)
```
Comment (tamird) [inline sql/parser/type_check.go]: ditto util.Errorf

**Comment by tamird** [comment]
LGTM modulo type safety

File: sql/parser/type_check.go (hunk 1 above)
Comment (petermattis) [inline sql/parser/type_check.go]: Not sure what you mean by statically. The grammar certainly never creates an `IF` expressions containing a different number of expressions. Would an `IfExpr` that implements `ConditionalExpr` only have slots for 3 expressions? Then we would need different code in `WalkExpr`.

File: sql/parser/eval.go (hunk 2 above)
Comment (petermattis) [inline sql/parser/eval.go]: Ok. Done.

File: sql/parser/type_check.go (hunk 3 above)
Comment (petermattis) [inline sql/parser/type_check.go]: Done.

File: sql/parser/type_check.go (hunk 1 above)
Comment (tamird) [inline sql/parser/type_check.go]: > Would an IfExpr that implements ConditionalExpr only have slots for 3 expressions?

Yeah, that's what I'm suggesting.

> Then we would need different code in WalkExpr.

I think that's OK, it wouldn't be much code.

If you made ConditionalExpr an interface, you could pull CaseExpr into that interface as well, which may allow you to share the typeCheck/evalExpr implementations

Comment (petermattis) [inline sql/parser/type_check.go]: Fitting `CaseExpr` into a `ConditionalExpr` interface for type-checking seems awkward as `CaseExpr` has quite a bit of different logic for handling the `when` conditions vs the `when` values. Let me give the `ConditionalExpr` interface a try for the existing conditionals.

Comment (petermattis) [inline sql/parser/type_check.go]: Added a commit which separates `ConditionalExpr` into `IfExpr`, `NullIfExpr` and `CoalesceExpr`. This made certain bits of the code cleaner, but also resulted in more code. I did not see an easy way to introduce a `ConditionalExpr` interface. Such an interface would have to have a function to return a list of the internal expressions, but then we'd need different code for type checking `If` vs `NullIf` vs `Coalesce`.

File: sql/parser/eval_test.go [hunk 4]
```diff
@@ -269,6 +269,19 @@ func TestEvalExpr(t *testing.T) {
...
+		{`IF(NULL, 1/0, 2)`, `2`},
+		{`NULLIF(1, 1)`, `NULL`},
+		{`NULLIF(1, 2)`, `1`},
+		{`NULLIF(2, 1)`, `2`},
+		{`IFNULL(1, 2/0)`, `1`},
+		{`IFNULL(NULL, 2)`, `2`},
+		{`IFNULL(1, NULL)`, `1`},
+		{`COALESCE(1, 2, 3, 4/0)`, `1`},
+		{`COALESCE(NULL, 2, 3, 4/0)`, `2`},
+		{`COALESCE(NULL, NULL, NULL, 4)`, `4`},
```
Comment (bdarnell) [inline sql/parser/eval_test.go]: Add a test for `coalesce` returning null (because all the arguments are null).

Comment (petermattis) [inline sql/parser/eval_test.go]: Oops. Added.

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
LGTM. Whoa, NULLIF is pretty weird

**Pull request merged**